vanniktech/gradle-maven-publish-plugin
# Real integration tests

Closes #42.

### What has been done
Configured integration tests (they are called functional on official Gradle site btw).

Tests added:
- [x] Java library
- [x] Java library + Groovy
- [ ] ~Android library~ - _will be added in_ #50
- [ ] ~Android library + Kotlin Android~ - _will be added in_ #50
- [ ] ~Java library + Kotlin~ - _will be added in_ #50

 
**TODO:**
- [x] Rewrite integration test from Groovy to Kotlin
- [x] Parametrize tests to make use of both `installArchives` and `uploadArchives` tasks
- [x] Open issue to add remaining tests listed in "Tests added" section above -> #50

### Integration tests execution time
By introducing integration tests, overall **CI build time arose by ~1 min**.
Compare build from this PR ([this one for example](https://travis-ci.org/vanniktech/gradle-maven-publish-plugin/builds/506753178)) with build from `master` branch ([this one for example](https://travis-ci.org/vanniktech/gradle-maven-publish-plugin/builds/502039095)).

### How to test

Add this dirty helper function somewhere after `installArchives`/`uploadArchives` command execution in test and execute tests via `./gradlew clean integrationTest --info` command to see generated artifacts in console.
```
  fun lsGeneratedFiles() {
    val p = Runtime.getRuntime().exec("ls -R ${repoFolder.absolutePath}/com/example/test-artifact")
    Scanner(p.inputStream).use { while (it.hasNextLine()) println(it.nextLine()) }
  }
```

#### Open question
> 💡 include `build.gradle` and `gradle.properties` files into fixture projects to ease debugging process or not? Ideally, these files should not be present and tests should construct required project configuration from scratch to have tests completely isolated. The only "downside" is that if test fails, it's easier to debug what went wrong, which artifacts were generated and which not, by running appropriate publish command from fixture project rather than from test itself (it's more natural to go to fixture project and run Gradle command). I would not include them, just raising a concern and it's relatively easy to construct `build.gradle` file using build file content from appropriate integration test.

NO!
